### PR TITLE
[MXNET-1107] Fix CPUPinned unexpected behaviour

### PR DESCRIPTION
**Pull request by ctcyang**
## Description ##
There is unexpected behavior with MXNet Context cpu_pinned for nonzero values of device.
This PR fixes this unexpected behavior.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix unexpected behavior with cpu_pinned Context
- [x] A class mxnet::common::cuda::SetDevice that collects what device was active before cudaSetDevice was called, and restores it to this device once the MXNet has done what it needs to do. This is done when the SetDevice object goes out of scope

## Comments ##
On master branch, if you do:
```
>>> import mxnet as mx
>>> data = mx.nd.zeros((2,3), ctx=mx.Context('cpu_pinned', 7))
>>> print(data)


[[0. 0. 0.]
 [0. 0. 0.]]
<NDArray 2x3 @cpu_pinned(7)>
```

This looks right. However, `nvidia-smi` tells me that this data is actually located on GPU 0:

```
$ nvidia-smi
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0    109140      C   python3                                      524MiB |
+-----------------------------------------------------------------------------+
```

To fix this unexpected behaviour, I changed four things:
1) call cudaSetDevice and set current device to GPU X we want to allocate on even for kCPUPinned case (src/storage/storage.cc)
2) before, Context::real_dev_id returns the GPU id for kGPU, and 0 for both kCPU and kCPUPinned even though it stores the dev_id that the user passed into the constructor as a private member (include/mxnet/base.h)
3) add guard to every place cudaSetDevice is called in MXNet C++ code. This guard is implemented using the class mxnet::common::cuda::SetDevice. The purpose of this guard is to prevent race conditions on which device was set first. Other GPU libraries have something similar e.g. 
https://github.com/uber/horovod/blob/master/horovod/torch/cuda_util.cc

### CODE REVIEW

File: include/mxnet/base.h [hunk 1]
```diff
@@ -156,7 +156,7 @@ struct Context {
    * \brief Returns dev_id for kGPU, 0 otherwise
```
Comment (haojin2) [inline include/mxnet/base.h]: Should also fix the documentation here?

Comment (ctcyang) [reply include/mxnet/base.h]: Thanks!

File: src/common/cuda_utils.h [hunk 2]
```diff
@@ -179,113 +283,38 @@ inline DType __device__ CudaMin(DType a, DType b) {
...
- * \param func Expression to call.
- *
- * It checks for cuRAND errors after invocation of the expression.
- */
-#define CURAND_CALL(func)                                       \
-  {                                                             \
-    curandStatus_t e = (func);                                  \
-    CHECK_EQ(e, CURAND_STATUS_SUCCESS)                          \
-        << "cuRAND: " << mxnet::common::cuda::CurandGetErrorString(e); \
+class SetDevice {
```
Comment (eric-haibin-lin) [inline src/common/cuda_utils.h]: Rename to DeviceScope, add a member function SetDevice, and remove all occurrences of `cudaGetDevice`?

**Comment by ctcyang** [comment]
Added [WIP] tag due to bug with `--kv-store dist_sync_device`.

**Comment by stu1130** [comment]
@mxnet-label-bot[pr-work-in-progress]

**Comment by pengzhao-intel** [comment]
@eric-haibin-lin Any update on this PR? 

It opens for a long time and we prefer to leverage the feature in this PR :)

File: Makefile [hunk 3]
```diff
@@ -678,3 +678,4 @@ clean_all: clean
 ifneq ($(EXTRA_OPERATORS),)
 	-include $(patsubst %, %/*.d, $(EXTRA_OPERATORS)) $(patsubst %, %/*/*.d, $(EXTRA_OPERATORS))
 endif
+
```
Comment (apeforest) [inline Makefile]: nit: remove blank line

File: src/kvstore/comm.h [hunk 4]
```diff
@@ -723,6 +723,9 @@ class CommDevice : public Comm {
     int n = static_cast<int>(gpus.size());
     int enabled = 0;
     std::vector<int> p2p(n*n);
+
+    // Restores active device to what it was before EnableP2P
+    mxnet::common::cuda::SetDevice set_device;
     for (int i = 0; i < n; ++i) {
       cudaSetDevice(gpus[i]);
```
Comment (apeforest) [inline src/kvstore/comm.h]: As @eric-haibin-lin suggested, replace this with a class method by your RAII object SetDevice (before renaming)

File: src/kvstore/comm_tree.h [hunk 5]
```diff
@@ -339,6 +339,7 @@ class CommDeviceTree : public CommDevice {
     int n = static_cast<int>(gpus.size());
     int enabled = 0;
     std::vector<int> p2p(n*n);
+    mxnet::common::cuda::SetDevice set_device;
     for (int i = 0; i < n; ++i) {
       cudaSetDevice(gpus[i]);
```
Comment (apeforest) [inline src/kvstore/comm_tree.h]: same here

File: src/kvstore/kvstore_nccl.h [hunk 6]
```diff
@@ -428,6 +428,7 @@ class KVStoreNCCL : public KVStoreLocal {
         mutate_vars.push_back(ptr(dst[i])->var());
     }
     Engine::Get()->PushSync([this](RunContext rctx) {
+        mxnet::common::cuda::SetDevice set_device;
         for (auto cur : nccl_data_) {
           CUDA_CALL(cudaSetDevice(cur.second.dev_id));
```
Comment (apeforest) [inline src/kvstore/kvstore_nccl.h]: same here

File: src/storage/storage.cc [hunk 7]
```diff
@@ -51,7 +51,13 @@ class StorageImpl : public Storage {
   static void ActivateDevice(Context ctx) {
```
Comment (apeforest) [inline src/storage/storage.cc]: Seems this method does not need to be static. It is only used in this class. Besides, making it static without a RAII class around is very dangerous.

Comment (apeforest) [inline src/storage/storage.cc]: This method should not be static. It is not protected by any RAII object. Besides, it is not used by any other classes anyway.

File: src/kvstore/comm.h (hunk 4 above)
Comment (apeforest) [inline src/kvstore/comm.h]: replace this with a class method by the SetDevice (renaming) object as @eric-haibin-lin suggested.

File: src/kvstore/comm_tree.h (hunk 5 above)
Comment (apeforest) [inline src/kvstore/comm_tree.h]: same here

File: src/kvstore/kvstore_nccl.h (hunk 6 above)
Comment (apeforest) [inline src/kvstore/kvstore_nccl.h]: same here

File: src/storage/storage.cc (hunk 7 above)
Comment (ctcyang) [reply src/storage/storage.cc]: You raise a good point, but I think that should be addressed in a separate PR.

File: Makefile (hunk 3 above)
Comment (ctcyang) [reply Makefile]: Fixed

**Pull request merged**